### PR TITLE
Add tristate to parent checkbox for DataTable

### PR DESCRIPTION
**Pull request by perclasson**
## Description

To follow Material specs, the data table parent checkbox should be a "tristate checkbox" and when it is pressed in the "intermediate" state it should select all: https://material.io/components/data-tables
data_table_checkboxes

## Related Issues

https://github.com/flutter/flutter/issues/64314

## Tests

I added the following tests:

- DataTable control test - tristate

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by rami-a — APPROVED** [review]

**Review by clocksmith — APPROVED** [review]

File: packages/flutter/lib/src/material/data_table.dart [hunk 1]
```diff
@@ -911,6 +919,7 @@ class DataTable extends StatelessWidget {
           onRowTap: () => row.onSelectChanged != null ? row.onSelectChanged(!row.selected) : null ,
           onCheckboxChanged: row.onSelectChanged,
           overlayColor: row.color ?? effectiveDataRowColor,
+          tristate: false,
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:922]: With NNBD we need to pass a value. Updated with master now.

File: packages/flutter/lib/src/material/data_table.dart [hunk 2]
```diff
@@ -847,6 +852,8 @@ class DataTable extends StatelessWidget {
     final bool anyRowSelectable = rows.any((DataRow row) => row.onSelectChanged != null);
     final bool displayCheckboxColumn = showCheckboxColumn && anyRowSelectable;
     final bool allChecked = displayCheckboxColumn && !rows.any((DataRow row) => row.onSelectChanged != null && !row.selected);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/data_table.dart:854]: If a table has tens of thousands of rows then scanning all of the rows 2X (worst case) seems like a bad idea.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/data_table.dart (hunk 2 above)
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart]: I optimized this now to:
```dart
final Iterable<DataRow> rowsChecked = displayCheckboxColumn ?
  rows.where((DataRow row) => row.onSelectChanged != null && row.selected) : <DataRow>[];
final bool allChecked = displayCheckboxColumn && rowsChecked.length == rows.length;
final bool anyChecked = displayCheckboxColumn && rowsChecked.isNotEmpty;
```

**Pull request merged**